boltons/ioutils.py

```diff
@@ -0,0 +1,322 @@
+# -*- coding: UTF-8 -*-
+
+# Coding decl above needed for rendering the emdash properly in the
+# documentation.
+
+"""
+Module ``ioutils`` implements a number of helper classes and functions which
+are useful when dealing with input, output, and bytestreams in a variety of
+ways.
+"""
+import os
+import sys
+from abc import (
+    ABCMeta,
+    abstractmethod,
+    abstractproperty,
+)
+from errno import EINVAL
+from io import BytesIO
+from codecs import EncodedFile
+from tempfile import TemporaryFile
+
+# Python2/3 compat
+if sys.version_info[0] == 3:
+    text_type = str
+    binary_type = bytes
+else:
+    text_type = unicode
+    binary_type = str
+
+
+def utf8_encode_if_unicode(string):
+    """
+    Given a string type, return it in bytes, encoding to UTF-8 if the string is
+    unicode.
+    """
+    if isinstance(string, text_type):
+        return string.encode('utf-8')
+    elif isinstance(string, binary_type):
+        return string
+    else:
+        raise TypeError("{} is not a unicode or str object".format(string))
+
+
+class SpooledIOBase(object):
+    """
+    The SpooledTempoaryFile class doesn't support a number of attributes and
+    methods that a StringIO instance does. This brings the api as close to
+    compatible as possible with StringIO so that it may be used as a near
+    drop-in replacement to save memory.
+
+    Another issue with SpooledTemporaryFile is that the spooled file is always
+    a cStringIO rather than a StringIO which causes issues with some of our
+    tools.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, max_size=5000000):
+        self._max_size = max_size
+
+    @abstractmethod
+    def read(self, n=-1):
+        """Read n characters from the buffer"""
+        pass
```

Comment:
I'll remove it; hopefully most of the people reading through it will know what's up.